File: WikipediaUnitTests/NSArray+WMFExtensions.m
//  Created by Monte Hurd on 8/11/15.
//  Copyright (c) 2015 Wikimedia Foundation. Provided under MIT-style license; please copy and modify!

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>
#import "Wikipedia-Swift.h"

@interface NSArray_WMFExtensions : XCTestCase

@end

@implementation NSArray_WMFExtensions

- (void)test_wmf_safeObjectAtIndex_findExpectedObject {
    NSArray* a = @[@"one", @"two"];
    XCTAssert([[a wmf_safeObjectAtIndex:0] isEqualToString:@"one"], @"Found expected object");
}

- (void)test_wmf_safeObjectAtIndex_outOfRangeReturnsNil {
    NSArray* a = @[@"one"];
    XCTAssertNil([a wmf_safeObjectAtIndex:1], @"Out of range returned nil properly.");
}

- (void)test_wmf_safeObjectAtIndex_emptyOutOfRangeReturnsNil {
    NSArray* a = @[];
    XCTAssertNil([a wmf_safeObjectAtIndex:1], @"Empty array returned nil properly.");
}

- (void)test_wmf_arrayByTrimmingToLength_countZeroReturnsSelf {
    NSArray* a = @[];
    XCTAssertEqualObjects(a, [a wmf_arrayByTrimmingToLength:5], @"Returned self on empty array.");
}

- (void)test_wmf_arrayByTrimmingToLength_arraySmallerThanRequestedLength {
    NSArray* a = @[@"one", @"two"];
    XCTAssertEqualObjects(a, [a wmf_arrayByTrimmingToLength:3], @"Returned self on too small length request.");
}

- (void)test_wmf_arrayByTrimmingToLength_trimToCount {
    NSArray* a  = @[@"one", @"two"];
    NSArray* a2 = [a wmf_arrayByTrimmingToLength:1];
    XCTAssert(a2.count == 1, @"Returned array of expected count.");
}

- (void)test_wmf_arrayByTrimmingToLength_trimToExpectedResult {
    NSArray* a = @[@"one", @"two"];
    XCTAssert([[a wmf_arrayByTrimmingToLength:1][0] isEqualToString:@"one"], @"Returned array containing expected object.");
}

- (void)test_wmf_reverseArray {
    NSArray* a  = @[@"one", @"two"];
    NSArray* a2 = [a wmf_reverseArray];

    XCTAssert([a2[0] isEqualToString:@"two"], @"Returned reversed array containing expected object.");
    XCTAssert([a2[1] isEqualToString:@"one"], @"Returned reversed array containing expected object.");
}

@end

File: Wikipedia/Categories/NSArray+WMFExtensions.swift
//  Created by Monte Hurd on 8/11/15.
//  Copyright (c) 2015 Wikimedia Foundation. Provided under MIT-style license; please copy and modify!

import Foundation

extension NSArray {
    /// @return The object at `index` if it's within range of the receiver, otherwise `nil`.
    public func wmf_safeObjectAtIndex(index: UInt) -> AnyObject? {
        let intIndex = Int(index);
        return (intIndex < self.count) ? self[intIndex] : nil;
    }

    /**
    Select up to `n` elements from an array.
    
    :param: length The max length
    
    :returns: A new array with the first `n` items in the receiver, or the receiver if `n` exceeds the number of items 
              in the array.
    */
    public func wmf_arrayByTrimmingToLength(n: UInt) -> NSArray {
        let intLength = Int(n);
        if (self.count == 0 || self.count < intLength) {
            return self;
        }
        return self.subarrayWithRange(NSMakeRange(0, intLength));
    }

    /**
    Select up to `n` elements from an array starting with the last element
    
    :param: length The max length
    
    :returns: A new array with the first `n` items in the receiver, or the receiver if `n` exceeds the number of items
    in the array.
    */
    public func wmf_arrayByTrimmingToLengthFromEnd(n: UInt) -> NSArray {
        let intLength = Int(n);
        if (self.count == 0 || self.count < intLength) {
            return self;
        }
        
        return self.subarrayWithRange(NSMakeRange(self.count-intLength, intLength));
    }

    /**
    Get all elements in an array except the first.

    :returns: All but the first element of the receiver, or an empty array if there was only one element.
    */
    public func wmf_arrayByRemovingFirstElement() -> NSArray {
        return wmf_safeSubarrayWithRange(NSMakeRange(1, self.count - 1))
    }

    /**
    Returns a subarray from the receiver, limited to its bounds.

    :param: range The range of the desired items.

    :returns: A subarray with the desired items, constrained by the number of items in the receiver.
    */
    public func wmf_safeSubarrayWithRange(range: NSRange) -> NSArray {
        if range.location > self.count - 1 || WMFRangeIsNotFoundOrEmpty(range) {
            return NSArray()
        }
        let safeLength: Int = {
            if WMFRangeGetMaxIndex(range) <= UInt(self.count) {
                return range.length
            } else {
                return max(0, self.count - range.location)
            }
        }()
        if safeLength == 0 {
            return NSArray()
        }
        precondition(safeLength >= 0)
        let safeRange = NSMakeRange(range.location, safeLength)
        precondition(WMFRangeGetMaxIndex(safeRange) <= UInt(self.count),
                     "Calculated unsafe range \(safeRange) for array of count \(self.count)")
        return self.subarrayWithRange(safeRange)
    }

    /// :returns: A reversed copy of the receiver.
    public func wmf_reverseArray() -> AnyObject {
        return self.reverseObjectEnumerator().allObjects;
    }
    
    /**
    Return an new array by interleaving the receiver with otherArray.
    The first element in the receiver is always first.
    
    :param: otherArray The array whose lements to interleave
    
    :returns: The interleaved array
    */
    public func wmf_arrayByInterleavingElementsFromArray(otherArray: NSArray) -> NSArray {
        
        var newArray = self.mutableCopy() as! NSMutableArray;
        
        otherArray.enumerateObjectsUsingBlock { (object, index, stop) -> Void in
            
            var newIndex = 2*index + 1;
            newIndex = newIndex > newArray.count ? newArray.count : newIndex;
            newArray.insertObject(object, atIndex: newIndex);
        }
        
        return newArray;
    }

}
